src/ucp/rndv/rndv.c

```diff
@@ -544,33 +547,51 @@ ucp_rndv_progress_rma_zcopy_common(ucp_request_t *req, ucp_lane_index_t lane,
     }
 }
 
+static UCS_F_ALWAYS_INLINE ucp_request_t*
+ucp_rndv_rma_get_zcopy_req(ucp_request_t *req);
+
 UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_rma_get_zcopy, (self),
                  uct_pending_req_t *self)
 {
-    ucp_request_t *rndv_req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucp_request_t *rndv_req;
     uct_rkey_t uct_rkey;
 
+    rndv_req = ucp_rndv_rma_get_zcopy_req(req);
+
     /* Figure out which lane to use for get operation */
-    rndv_req->send.lane =
-        ucp_rndv_zcopy_get_lane(rndv_req, &uct_rkey,
-                                UCP_REQUEST_SEND_PROTO_RNDV_GET);
-    if (rndv_req->send.lane != UCP_NULL_LANE) {
+    req->send.lane =
+            ucp_rndv_zcopy_get_lane(req, &uct_rkey,
+                                    UCP_REQUEST_SEND_PROTO_RNDV_GET);
+    if (req->send.lane != UCP_NULL_LANE) {
         return ucp_rndv_progress_rma_zcopy_common(
-                rndv_req, rndv_req->send.lane, uct_rkey,
+                req, req->send.lane, uct_rkey,
                 UCP_REQUEST_SEND_PROTO_RNDV_GET);
     }
 
+    /* GET Zcopy to staging buffers part of PUT pipeline shouldn't go here */
+    ucs_assert(rndv_req->send.rndv.rkey != NULL);
+
     /* If can't perform get_zcopy - switch to active-message.
      * NOTE: we do not register memory and do not send our keys. */
     ucp_trace_req(rndv_req, "remote memory unreachable, switch to rtr");
     ucp_rkey_destroy(rndv_req->send.rndv.rkey);
     ucp_rndv_recv_data_init(ucp_request_get_super(rndv_req),
                             rndv_req->send.length);
     /* Remove statistics related to get_zcopy */
-    UCP_WORKER_STAT_RNDV(rndv_req->send.ep->worker, GET_ZCOPY, -1);
+    UCP_WORKER_STAT_RNDV(req->send.ep->worker, GET_ZCOPY, -1);
+
+    if (req != rndv_req) {
+        /* If GET Zcopy operation was issued as a part of fragmented protocol,
+         * release the fragmented request which is no longer needed */
+        ucs_mpool_put_inline((void*)req->send.mdesc);
+        ucp_request_put(req);
```

Comment:
can we share code with get-frag completion?

Reply:
unfortunately, no.
it could not be shared with put completion that's called in the end of GET fragmented